docs/dsl/example.md

```diff
@@ -0,0 +1,438 @@
+
+The nodes are Python classes and the arrows are methods on those classes.
+
+```
+             +---+
+             |   V
+      filter |  PatientFrame
+             |   |   |
+             +---+   | select_column
+                     V
+                PatientSeries
+
+
+             +---+
+             |   V
+      filter |  EventFrame -----------------------+
+             |   |   |                            |
+             +---+   | sort_by                    |
+                     V                            |
+             SortedEventFrame                     |
+                     |                            |
+        +---+        | (first/last)_for_patient   | (count/exists)_for_patient
+        |   V        V                            |
+ filter |  AggregatedEventFrame                   |
+        |   |        |                            |
+        +---+        | select_column              |
+                     V                            |
+               PatientSeries <--------------------+
+```
+
+---
+
+Here's the code that we can use to generate an input.csv table.
+
+```
+from cohortextractor2.concepts.tables import (
+    immunisations as imms,
+    patients,
+    registrations,
+)
+from cohortextractor2.dsl import Cohort, categorise
+
+from codelists import covid_vacc_codes
+
+# Create a cohort.
+cohort = Cohort()
+
+# Create two PatientSeries for age and sex and assign to the cohort.
+cohort.age = patients.select_column(patients.age)
+cohort.sex = patients.select_column(patients.sex)
+
+# Create a new EventFrame representing just the events with covid vaccination codes.
+covid_imms = imms.filter(covid_vacc_codes.contains(imms.code))
+
+# Create an AggregatedEventFrame representing just, for each patient, the first record
+# in covid_imms.
+first = covid_imms.sort_by(imms.date).first_for_patient()
+
+# Create two PatientSeries for the first date / code and assign to the cohort.
+cohort.first_date = first.select_column(imms.date)
+cohort.first_code = first.select_column(imms.code)
+
+# Create an AggregatedEventFrame representing, for each patient, just the record that
+# occurs first 28 days after the first record of a covid vaccination.
+second = (
+    covid_imms.filter(imms.date > (cohort.first_date + 28))
+    .sort_by(imms.date)
+    .first_for_patient()
+)
+
+# Create two PatientSeries for the second date / code and assign to the cohort.
+cohort.second_date = second.select_column("date")
+cohort.second_code = second.select_column("code")
+
+# Create a PatientSeries based on the values in two other PatientSeries.
+cohort.codes_match = categorise(
+    {
+        "yes": cohort.first_code == cohort.second_code,
+        "no": True,
+    }
+)
+
+# Include only patients who have been at same practice for duration of study.
+cohort.set_population(
+    registrations.filter(registrations.start_date <= "2021-01-01")
+    .filter(registrations.end_date >= "2021-06-30")
+    .exists_for_patient()
+)
+```
+
```

Comment:
thought: In the video demo, `get_cohort_variables()` was used to demonstrate what the query model looks like.

It might be worth mentioning, perhaps labelled as an "implementation detail" point not required to understand the DSL, that:

* the DSL gets transformed to an internal query model, which then gets dispatched to a backend-specific query
* the query model representation can be accessed via `get_cohort_variables()` 
* an example of what the output of `get_cohort_variables()` looks like